Dockerfile

```diff
@@ -1,29 +1,115 @@
-FROM ruby:2.7.2
-
-# 1. Configuring apt to properly install yarn
-# 2. Installing node, postgres-client, yarn
-# 3. Clearing apt-cached files to tighten the image size
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1 apt-key add - \
-  && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list \
-  && apt-get update -qq \
-  && apt-get install -y build-essential \
-  nodejs \
-  yarn \
-  && rm -rf /var/lib/apt/lists/*
+# Generating a generic Rails image.
+# Just for caching.
+FROM ruby:2.7-alpine AS rails
 
-WORKDIR /app
+# `apk` is like `apt`, but for Alpine Linux.
+RUN apk add --update --no-cache --quiet \
+    build-base \
+    nodejs-current \
+    postgresql-dev \
+    sqlite-dev \
+    tzdata \
+    yarn
+
+# Create a generic Gemfile with only Rails.
+RUN echo -e "source 'https://rubygems.org'\ngem 'rails'" > /Gemfile
+# Install Rails, quietly, in 4 threads.
+RUN bundle install --quiet -j4
+
+# Uses the image above as the base for a builder image.
+# This will **exclude** development and testing gems.
+FROM rails as rails-builder
+
+# Set the value of `pwd` for all following operations.
+WORKDIR /app/
+
+# Add the Gemfile and Gemfile.lock from this application.
+COPY Gemfile Gemfile.lock /app/
+
+# Installs gems, quietly, in 4 threads, and retry if broken.
+RUN bundle install --quiet -j4 --retry 3 \
+    # Remove unneeded files (cached *.gem, *.o, *.c)
+    && rm -rf /usr/local/bundle/cache/*.gem \
+    && find /usr/local/bundle/gems/ -name "*.c" -delete \
+    && find /usr/local/bundle/gems/ -name "*.o" -delete
```

Comment:
Why do we need to do this? Doesn't bundle handle this for us?

Reply:
Bundle doesn't clean up headers and other files when compiling nokogiri or other gems with native implementations written in C.  This does.  

The intent is to shrink the size of the image by removing files not needed in Prod.